xarray/backends/zarr.py

```diff
@@ -0,0 +1,516 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import functools
+import warnings
+from itertools import product
+from collections import MutableMapping
+import operator
+
+import numpy as np
+
+from .. import Variable
+from ..core import indexing
+from ..core.utils import (FrozenOrderedDict, close_on_error, HiddenKeyDict,
+                          NdimSizeLenMixin,DunderArrayMixin)
+from ..core.pycompat import (iteritems, bytes_type, unicode_type, OrderedDict,
+                             basestring)
+
+from .common import (WritableCFDataStore, AbstractWritableDataStore,
+                     DataStorePickleMixin)
+
+from .. import conventions
+
+# most of the other stores have some kind of wrapper class like
+# class BaseNetCDF4Array(NdimSizeLenMixin, DunderArrayMixin):
+# class H5NetCDFArrayWrapper(BaseNetCDF4Array):
+# class NioArrayWrapper(NdimSizeLenMixin, DunderArrayMixin):
+# we problaby need something like this
+
+# the first question is whether it should be based on BaseNetCDF4Array or
+# NdimSizeLenMixing?
+
+# or maybe we don't need wrappers at all? probably not true
+
+# zarr attributes have to be serializable as json
+# many xarray datasets / variables have numpy arrays and values
+# these functions handle encoding / decoding of such items
+def _encode_zarr_attr_value(value):
+    # what is the most duck-type friendly way to do this check
+    print("encoding zarr attr value %r" % value)
+    if isinstance(value, np.ndarray):
+        return value.tolist()
+    # I don't know how to check generically if something is a numpy scalar
+    # i.e. np.float32 or np.int8, etc. without checking against each dtype
+    # manually. This was the best I could come up with
+    try:
+        # np.string_('X').item() returns a type `bytes`
+        # zarr still doesn't like that
+        # causes some fill_value encoding to fail
+        return value.item()
+    except AttributeError:
+        return value
```

Comment:
@shoyer: this is my attempt to encode numpy type attributes for zarr. It works for many cases but still fails when the data is a `bytes` type (with error `TypeError: Object of type 'bytes' is not JSON serializable`).

Is it safe to always convert `bytes` to strings? What else might bytes be used for? Is there a more generic way to encode bytes?

Reply:
> Is it safe to always convert bytes to strings? What else might bytes be used for?

Sadly, no. `bytes` can represent arbitrary binary data. For example, if you call `ds.to_netcdf()` you get `bytes` back corresponding to a netCDF3 file in memory. In theory, you could put that in an attribute (of course, this would be a terrible idea). In practice, most attributes that are `bytes` as ASCII characters.

> Is there a more generic way to encode bytes?

Basically, we have two options:

1. Most strict: only allow bytes that are valid ASCII characters (only the values 0-127, excluding 128-255). Raise an error for everything else. To do this, we would convert using `value.decode('ascii')`, which will raise `UnicodeDecodeError` when invalid characters are encountered (we would want to wrap this error message to clarify where it was encountered).
2. Most flexible: Use something like [`base64.b64encode`](https://docs.python.org/3.6/library/base64.html) to convert bytes into unicode.

Either way, I would only do this for specific attributes on which we can reverse the encoding, e.g., `fill_value` if `dtype == np.string_`. Doing this for arbitrary attributes is probably not a good idea, since we won't know when to encode back.